### PR TITLE
[PR] Creating Flutter Todo list with Bloc

### LINKED ISSUE #1: Why? What? Who? How?
# Why?

@LuchoTurtle recently opened the following issue: https://github.com/dwyl/app/issues/320
We need to build an example `App` akin to our `MVP`: https://github.com/dwyl/mvp 
We cannot simply pick a technology / architecture based on a recommendation from someone we know.
We _must_ make the effort up-front to document our choice with a clear step-by-step example
that everyone in the team can use to get up-to-speed.

Once that is done we can evaluate if `bloc` is a good fit for our use-case. 

# What?

+ [x] Briefly Explain the `BLoC` Architecture 
+ [x] Explain how the `bloc` library works and why it's a good fit for medium to larger `Flutter` Apps. 

# Who?

This tutorial is targeted at the ***novice*** `Flutter` dev. 
We are writing a step-by-step tutorial as an authoritative guide to the `@dwyl` team/community up-to-speed.
If we can make this tutorial useful to other people in the wider `Flutter` community that's great!

# How?

The expectation is that they have already followed at least 2 of our other tutorials:
1. https://github.com/dwyl/learn-flutter
2. https://github.com/dwyl/flutter-todo-list-tutorial

These should be included as prerequisites in the `README.md` "How?" section 

# Todo

+ [x] Follow test-driven development from the start. 
  + [x] Setup the project and testing in the **_first_ step** of the tutorial. 
  + [x] Setup `GitHub Actions CI` and track coverage for the project. 
    + [x] Project should have 100% coverage when complete. 

+ [x] Follow and _combine_ the following tutorials:
  + [x] Timer: https://bloclibrary.dev/#/fluttertimertutorial
  + [x] Todo List: https://bloclibrary.dev/#/fluttertodostutorial
  + [x] Showcase how `bloc` helps to separate presentation from business logic.
+ [x] Include as many in-line comments in the code as would be reasonable for a _beginner_ who has never seen `bloc` before to understand what is going on. i.e. the code is not "self documenting" ...

> **Note**: for now, it's OK for all data to remain on the device. i.e. no need to interact with our `API`.
But if you're feeling like you want to _extend_ the tutorial, feel free to dive into saving data to the `API`. 👌 

+ [x] Bonus Level: Make it look great! Ideally try to match the UI in our Figma for the @dwyl app: https://www.figma.com/file/WrpkKGJNYZbeCzMRDVtvQLbC/dwyl-app related to: https://github.com/dwyl/app/issues/328

### PR DESCRIPTION
**Pull request by LuchoTurtle**
closes #1 

Still very much in-progress with some local changes still unpushed.

### CODE REVIEW

**Comment by nelsonic** [comment]
Looking good so far. Can't wait to read the write-up. 👌 
Safe travels. ✈️

**Comment by LuchoTurtle** [comment]
@nelsonic this should be good for review.
I'll continue working on making the UI as close to the Figma files as possible. 👍

**Comment by nelsonic** [comment]
Need to time-box `T2h` to review this. ⏳

File: README.md [hunk 1]
```diff
@@ -0,0 +1,2469 @@
...
+we need to:
+- provide an initial state.
+- set up event listeners and handlers.
+- add events to `bloc` via `bloc.add`.
+
+Check the following piece of code.
+
+```dart
+class PetBloc extends Bloc<PetEvent, PetState> {
+    final PetRepo repo;
```
Comment (nelsonic) [inline README.md:360]: what is the `repo`? 
(please add a comment linking docs to where people can understand this boilerplate code ... 🔗 🙏 )
e.g: https://bloclibrary.dev/#/architecture?id=repository
But documentation uses "`repository`" ... looks like it's just a variable name. 
maybe we should use the same one for consistency. 💭

File: README.md [hunk 2]
```diff
@@ -0,0 +1,2469 @@
...
+
+    // Defining the class
+    PetBloc(this.repo)) : super(PetInitial()) {
+        on<LoadPet>(_loadPet);
```
Comment (nelsonic) [inline README.md:364]: why are the event handlers prefixed with an underscore? `_loadPet` 🤷‍♂️ 
if there isn't a _good_ reason for using (_needing_) them, can we please _not_ use them as they are confusing
for anyone who comes from a language where `_` means "ignore" this. 💭

**Review by nelsonic — APPROVED** [review]
@LuchoTurtle very good starting point. 🚀 
I would change a few more things ... 💭 
But prefer to merge this as-is so that we can start _using_ it. :shipit: 

image

Thanks! 🙏

File: README.md (hunk 2 above)
Comment (LuchoTurtle) [reply README.md:364]: [Because they're private.](https://stackoverflow.com/questions/53142171/what-does-underscore-before-variable-name-mean-for-flutter)

Comment (nelsonic) [reply README.md:364]: Cool. Knew there was something. But feels like something Dart should have a better symbol for. 💭

**Pull request merged**